File: backend/models/schemas.py
from pydantic import BaseModel, Field
from typing import Literal, Optional
from enum import Enum


# --- Enums ---
class Seniority(str, Enum):
    junior = "junior"
    mid = "mid"
    senior = "senior"
    staff = "staff"


class TrendDirection(str, Enum):
    rising = "rising"
    stable = "stable"
    declining = "declining"


class HeatmapColor(str, Enum):
    green = "green"
    yellow = "yellow"
    red = "red"


# --- Profile ---
class ManualProfileInput(BaseModel):
    skills: list[str] = Field(..., min_length=1, max_length=20)       
    seniority: Seniority
    location: str = Field(..., min_length=2, max_length=50)
    years_experience: int = Field(..., ge=0, le=50)
    current_role: str = Field(..., min_length=2, max_length=100)      


class GitHubProfileInput(BaseModel):
    github_url: str = Field(..., pattern=r'^https://github\.com/[a-zA-Z0-9_-]+$')


class LanguageBreakdown(BaseModel):
    language: str
    percentage: float
    repos_count: int


class GitHubProfileOutput(BaseModel):
    username: str
    detected_skills: list[str]
    primary_languages: list[LanguageBreakdown]
    estimated_seniority: Seniority
    years_active: int
    total_repos: int
    total_commits_last_year: int
    notable_patterns: list[str]
    profile_summary: str


# --- Diagnosis ---
class SalaryRange(BaseModel):
    min: int
    max: int
    currency: str = "USD"


class SalaryDiagnosis(BaseModel):
    current_range: SalaryRange
    potential_range: SalaryRange
    gap_annual: int
    gap_3year: int
    current_job_count: int
    potential_job_count: int
    job_count_increase_pct: int
    key_missing_skill: str
    secondary_missing_skills: list[str]


class MarketScore(BaseModel):
    overall: int = Field(..., ge=0, le=100)
    breakdown: dict[str, int]  # e.g. {"skill_demand": 78, "skill_breadth": 65, ...}


class SkillAnalysis(BaseModel):
    skill_name: str
    current_level: str
    market_demand: TrendDirection
    percentile_rank: int = Field(..., ge=0, le=100)
    salary_impact_usd: int
    trend_direction: TrendDirection
    trend_growth_pct: float


class PeerComparison(BaseModel):
    axes: list[str]
    user_values: list[int]
    peer_avg_values: list[int]
    seniority_group: str
    region: str
    overall_percentile: int
    percentile_label: str


class Opportunity(BaseModel):
    skill: str
    unlock_count: int
    salary_increase_pct: int
    salary_increase_usd: int
    demand_trend: TrendDirection
    trend_growth_pct: float
    difficulty: Literal["low", "medium", "high"]
    time_to_learn: str


class HeatmapEntry(BaseModel):
    skill: str
    trend: TrendDirection
    growth_pct: float
    color: HeatmapColor


class DiagnosisResponse(BaseModel):
    salary_diagnosis: SalaryDiagnosis
    market_score: MarketScore
    skills_analysis: list[SkillAnalysis]
    peer_comparison: PeerComparison
    opportunities: list[Opportunity]
    demand_heatmap: list[HeatmapEntry]
    market_summary: str
    value_narrative: str


# --- What-If ---
class WhatIfRequest(BaseModel):
    current_skills: list[str]
    hypothetical_add: list[str] = []
    hypothetical_remove: list[str] = []
    seniority: Seniority
    location: str


class WhatIfResponse(BaseModel):
    new_score: int
    score_change: str
    new_salary_range: SalaryRange
    salary_change_usd: int
    new_job_count: int
    job_count_change: str
    new_percentile: int
    percentile_change: str
    insight: str


# --- Negotiation ---
class ChatMessage(BaseModel):
    role: Literal["user", "assistant"]
    content: str


class NegotiateRequest(BaseModel):
    company: Literal["mercadolibre", "globant", "nubank", "rappi"]    
    role: str
    user_profile: ManualProfileInput
    conversation_history: list[ChatMessage] = []
    user_message: str


class ArgumentImpact(BaseModel):
    argument: str
    impact_usd: int
    accepted: bool


class NegotiateResponse(BaseModel):
    ai_response: str
    current_offer: int
    turn_number: int
    negotiation_complete: bool
    argument_impact: Optional[ArgumentImpact] = None


# --- Negotiation Report ---
class WorkedArgument(BaseModel):
    argument: str
    impact_usd: int


class FailedArgument(BaseModel):
    argument: str
    reason: str


class SkillGap(BaseModel):
    skill: str
    impact_usd: int


class NegotiationReportRequest(BaseModel):
    company: Literal["mercadolibre", "globant", "nubank", "rappi"]    
    role: str
    user_profile: ManualProfileInput
    full_conversation: list[ChatMessage]
    final_offer: int
    initial_offer: int


class NegotiationReportResponse(BaseModel):
    final_offer: int
    initial_offer: int
    negotiated_increase: int
    what_worked: list[WorkedArgument]
    what_didnt_work: list[FailedArgument]
    current_ceiling: int
    potential_ceiling: int
    skills_to_close_gap: list[SkillGap]


# --- Roadmap ---
class RoadmapPhase(BaseModel):
    phase: str
    actions: list[str]


class RoadmapResponse(BaseModel):
    target_skill: str
    impact_usd: int
    total_duration: str
    phases: list[RoadmapPhase]

File: frontend/lib/schemas.ts
import { z } from "zod";

const SeniorityEnum = z.enum(["junior", "mid", "senior", "staff"]);
const TrendEnum = z.enum(["rising", "stable", "declining"]);
const ColorEnum = z.enum(["green", "yellow", "red"]);
const CompanyEnum = z.enum(["mercadolibre", "globant", "nubank", "rappi"]);

export const ManualProfileSchema = z.object({
  skills: z.array(z.string()).min(1).max(20),
  seniority: SeniorityEnum,
  location: z.string().min(2).max(50),
  years_experience: z.number().int().min(0).max(50),
  current_role: z.string().min(2).max(100),
});

export const GitHubProfileInputSchema = z.object({
  github_url: z.string().url().optional(),
  github_username: z.string().min(1).optional(),
});

export const GitHubProfileRequestSchema = GitHubProfileInputSchema.refine(
  (value) => Boolean(value.github_url || value.github_username),
  {
    message: "github_username or github_url is required",
    path: ["github_username"],
  },
);

export const ConversationMessageSchema = z.object({
  role: z.enum(["user", "assistant"]),
  content: z.string().min(1),
});

export const GitHubProfileOutputSchema = z.object({
  username: z.string(),
  detected_skills: z.array(z.string()),
  primary_languages: z.array(
    z.object({
      language: z.string(),
      percentage: z.number(),
      repos_count: z.number(),
    }),
  ),
  estimated_seniority: SeniorityEnum,
  years_active: z.number(),
  total_repos: z.number(),
  total_commits_last_year: z.number(),
  notable_patterns: z.array(z.string()),
  profile_summary: z.string(),
});

const SalaryRangeSchema = z.object({
  min: z.number(),
  max: z.number(),
  currency: z.string().default("USD"),
});

export const DiagnosisResponseSchema = z.object({
  salary_diagnosis: z.object({
    current_range: SalaryRangeSchema,
    potential_range: SalaryRangeSchema,
    gap_annual: z.number(),
    gap_3year: z.number(),
    current_job_count: z.number(),
    potential_job_count: z.number(),
    job_count_increase_pct: z.number(),
    key_missing_skill: z.string(),
    secondary_missing_skills: z.array(z.string()),
  }),
  market_score: z.object({
    overall: z.number().min(0).max(100),
    breakdown: z.record(z.string(), z.number()),
  }),
  skills_analysis: z.array(
    z.object({
      skill_name: z.string(),
      current_level: z.string(),
      market_demand: TrendEnum,
      percentile_rank: z.number(),
      salary_impact_usd: z.number(),
      trend_direction: TrendEnum,
      trend_growth_pct: z.number(),
    }),
  ),
  peer_comparison: z.object({
    axes: z.array(z.string()),
    user_values: z.array(z.number()),
    peer_avg_values: z.array(z.number()),
    seniority_group: z.string(),
    region: z.string(),
    overall_percentile: z.number(),
    percentile_label: z.string(),
  }),
  opportunities: z.array(
    z.object({
      skill: z.string(),
      unlock_count: z.number(),
      salary_increase_pct: z.number(),
      salary_increase_usd: z.number(),
      demand_trend: TrendEnum,
      trend_growth_pct: z.number(),
      difficulty: z.enum(["low", "medium", "high"]),
      time_to_learn: z.string(),
    }),
  ),
  demand_heatmap: z.array(
    z.object({
      skill: z.string(),
      trend: TrendEnum,
      growth_pct: z.number(),
      color: ColorEnum,
    }),
  ),
  market_summary: z.string(),
  value_narrative: z.string(),
});

export const WhatIfRequestSchema = z.object({
  current_skills: z.array(z.string()).max(30),
  hypothetical_add: z.array(z.string()).max(10),
  hypothetical_remove: z.array(z.string()).max(10),
  seniority: SeniorityEnum,
  location: z.string().min(2).max(50),
});

export const WhatIfResponseSchema = z.object({
  new_score: z.number(),
  score_change: z.string(),
  new_salary_range: SalaryRangeSchema,
  salary_change_usd: z.number(),
  new_job_count: z.number(),
  job_count_change: z.string(),
  new_percentile: z.number(),
  percentile_change: z.string(),
  insight: z.string(),
});

export const NegotiateRequestSchema = z.object({
  company: CompanyEnum,
  role: z.string().min(2).max(120),
  user_profile: z.record(z.string(), z.unknown()),
  conversation_history: z.array(ConversationMessageSchema).max(30),
  user_message: z.string().min(1).max(2000),
});

export const NegotiateResponseSchema = z.object({
  ai_response: z.string(),
  current_offer: z.number(),
  turn_number: z.number(),
  negotiation_complete: z.boolean(),
  argument_impact: z
    .object({
      argument: z.string(),
      impact_usd: z.number(),
      accepted: z.boolean(),
    })
    .nullable(),
});

export const NegotiationReportSchema = z.object({
  final_offer: z.number(),
  initial_offer: z.number(),
  negotiated_increase: z.number(),
  what_worked: z.array(
    z.object({
      argument: z.string(),
      impact_usd: z.number(),
    }),
  ),
  what_didnt_work: z.array(
    z.object({
      argument: z.string(),
      reason: z.string(),
    }),
  ),
  current_ceiling: z.number(),
  potential_ceiling: z.number(),
  skills_to_close_gap: z.array(
    z.object({
      skill: z.string(),
      impact_usd: z.number(),
    }),
  ),
});

export const NegotiationReportRequestSchema = z.object({
  company: CompanyEnum.optional(),
  role: z.string().min(2).max(120).optional(),
  user_profile: z.record(z.string(), z.unknown()).optional(),
  full_conversation: z.array(ConversationMessageSchema).optional(),
  final_offer: z.number().positive(),
  initial_offer: z.number().positive(),
});
